### ISSUE TITLE
bump-formula-pr should update SHA256 for resources

### ISSUE THREAD

**Issue by nicholasdille** [Homebrew/brew#10620]
**Provide a detailed description of the proposed feature**

Whem bump-formula-pr replaced the old version with the new version this also applies to resource blocks but the corresponding SHA256 is not updated.

I think that bump-formula-pr should update the SHA256 if the corresponding URL was modified.

**What is the motivation for the feature?**

The changes in bump-formula-pr would allow formulas to be upgraded automatically. I am thinking about the following workflow:

1. Use renovate bot (or a similar tool) to find a new version

1. Use `bump-formula-pr` to create a PR on my tap

1. Review the results of the tests run by the self-hosted GitHub Action

**How will the feature be relevant to at least 90% of Homebrew users?**

It will only indirectly affect Homebrew users and certainly not in a way that many users will notice.

**What alternatives to the feature have been considered?**

1. Although renovate supports Homebrew, it assumes that a URL points to the tarball of the corresponding branch

1. Some ruby code could update SHA256 values after `bump-formula-pr` which feels ugly

**Comment by Rylan12** [comment]
Maybe I'm misunderstanding a little, but why are your resource blocks being updated by `bump-formula-pr`? My understanding is that the only time `bump-formula-pr` will modify resource blocks is for python formulae with PyPI resources that can be automatically updated. When doing this, the `sha256` methods are updated appropriately.

Can you give an example of a formula and `bump-formula-pr` command where this issue occurs?

**Comment by nicholasdille** [comment]
I have a binary formula that installs `regctl` with two companion tools. All three are distributed as separate assets in a Github release. My formula installs them together:

https://github.com/nicholasdille/homebrew-tap/blob/main/Formula/regclient-bin.rb

This formula is probably an edge case and makes more sense when compiling from source 😉

You can reproduce this by bumping the formula:

1. `brew tap nicholasdille/tap`

1. `brew bump-formula-pr --version=0.2.0 --strict --no-browse --no-fork --write nicholasdille/tap/regclient-bin`

As a novice to Homebrew, it feels consistent to update URL and sha256 of resources.

**Comment by Rylan12** [comment]
Ah, I see. Could those resources be other formulae that the `regclient-bin` formula depends on?

In general, `resources` aren't really meant to be updated on each formula upgrade, if you know what I mean. In general, if a resource needs to be regularly updated it might make more sense to have it be its own formula (the exception to this is seems to be PyPI packages).

I'd consider adding an e.g. `--update-resource-checksums` flag to `bump-formula-pr` that would recalculate and write new checksums for resources. Then, you could use something like this to ensure that the resource always stays in sync with the main version.

```ruby
  resource "regsync" do
    url "https://github.com/regclient/regclient/releases/download/v#{version}/regsync-linux-amd64"
    sha256 "18fa2f3cc439297685ed69fd6c7506843bdbd960a176451efe7f4765e0130626"
  end
  resource "regbot" do
    url "https://github.com/regclient/regclient/releases/download/v#{version}/regbot-linux-amd64"
    sha256 "3670c90861647c1074bfb593cd264e20d47377a68ca22beeaec2fe1ee5f1a15b"
  end
  resource "docker-plugin" do
    url "https://github.com/regclient/regclient/raw/v#{version}/docker-plugin/docker-regclient"
    sha256 "e04c8fc0d634d20cda6185f425a214d1c5246e94566311d7bc6caff03accb302"
  end
```

Of course, this would break if the resource URL scheme changes or the version numbers go out of sync, but that would be easily fixable and may never happen if you're in control of the repos.

Another option would be to write your own command for updating the resources/formulae. `bump-formula-pr` is certainly nice, but it needs to be extremely flexible to fit Homebrew's internal needs. A custom command could be better tailored for your specific situation as you can make assumptions about things we won't make (e.g. version bumps _always_ come with resource bumps and the version _always_ matches).

**Comment by nicholasdille** [comment]
I have given this some thought. This feature request only affects the binary formulae I have created. I think I'd rather remove them in favour of bottled formulae.

I am very happy with the functionality provided by `bump-formula-pr` so I'd rather not create additional tooling as long as it can be avoided.

I was just confused that the version got updated in resource URLs but the SHA256 remained untouched.
